File: mobile/src/main/java/info/papdt/express/helper/dao/PackageDatabase.java
package info.papdt.express.helper.dao;

import android.content.Context;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.annotations.Expose;

import java.io.IOException;
import java.util.ArrayList;

import info.papdt.express.helper.api.PackageApi;
import info.papdt.express.helper.model.BaseMessage;
import info.papdt.express.helper.model.Package;
import info.papdt.express.helper.support.FileUtils;

public class PackageDatabase {

	@Expose private ArrayList<Package> data;
	private ArrayList<Package> delivered, delivering;
	private Context mContext;

	private static PackageDatabase sInstance;

	private Package lastRemovedData = null;
	private int lastRemovedPosition = -1;

	private static final String FILE_NAME = "packages.json";

	private static final String TAG = PackageDatabase.class.getSimpleName();

	public static PackageDatabase getInstance(Context context) {
		if (sInstance == null) {
			sInstance = new PackageDatabase(context);
		}
		return sInstance;
	}

	private PackageDatabase(Context context) {
		this.mContext = context;
		this.load();
	}

	public void load() {
		String json;
		try {
			json = FileUtils.readFile(mContext, FILE_NAME);
		} catch (IOException e) {
			json = "{\"data\":[]}";
			e.printStackTrace();
		}
		this.data = new Gson().fromJson(json, PackageDatabase.class).data;
		refreshList();
	}

	public boolean save() {
		try {
			Gson gson = new GsonBuilder().excludeFieldsWithoutExposeAnnotation().create();
			FileUtils.saveFile(mContext, FILE_NAME, gson.toJson(this));
			return true;
		} catch (IOException e) {
			e.printStackTrace();
			return false;
		}
	}

	public void refreshList() {
		delivered = new ArrayList<>();
		delivering = new ArrayList<>();
		for (Package p : data) {
			if (p.getState() == Package.STATUS_DELIVERED) {
				delivered.add(p);
			} else {
				delivering.add(p);
			}
		}
	}

	public void add(Package pack) {
		data.add(pack);
		refreshList();
	}

	public void add(int index, Package pack) {
		data.add(index, pack);
		refreshList();
	}

	public void set(int index, Package pack) {
		data.set(index, pack);
		refreshList();
	}

	public void remove(int index) {
		final Package removedItem = data.remove(index);

		lastRemovedData = removedItem;
		lastRemovedPosition = index;

		refreshList();
	}

	public void remove(Package pack) {
		int nowPos = indexOf(pack);
		remove(nowPos);
	}

	public int undoLastRemoval() {
		if (lastRemovedData != null) {
			int insertedPosition;
			if (lastRemovedPosition >= 0 && lastRemovedPosition < data.size()) {
				insertedPosition = lastRemovedPosition;
			} else {
				insertedPosition = data.size();
			}

			data.add(insertedPosition, lastRemovedData);
			refreshList();

			lastRemovedData = null;
			lastRemovedPosition = -1;

			return insertedPosition;
		} else {
			return -1;
		}
	}

	public ArrayList<Package> getData() {
		return data;
	}

	public ArrayList<Package> getDeliveredData() {
		return delivered;
	}

	public ArrayList<Package> getDeliveringData() {
		return delivering;
	}

	public int indexOf(Package p) {
		return indexOf(p.number);
	}

	public int indexOf(String number) {
		for (int index = 0; index < size(); index++) {
			if (get(index).number.equals(number)) return index;
		}
		return -1;
	}

	public void clear() {
		data.clear();
		refreshList();
	}

	public int size() {
		return data.size();
	}

	public Package get(int index) {
		return data.get(index);
	}

	public void pullDataFromNetwork(boolean shouldRefreshDelivered) {
		for (int i = 0; i < size(); i++) {
			Package pack = this.get(i);
			if (!shouldRefreshDelivered && pack.getState() == Package.STATUS_DELIVERED) {
				continue;
			}
			BaseMessage<Package> newPack = PackageApi.getPackage(pack.companyType, pack.number);
			if (newPack.getCode() == BaseMessage.CODE_OKAY && newPack.getData().data != null) {
				newPack.getData().shouldPush = newPack.getData().data.size() > pack.data.size();
				this.set(i, newPack.getData());
			} else {
				Log.e(TAG, "Package " + pack.codeNumber + " couldn\'t get new info.");
			}
		}
		refreshList();
	}

}

File: mobile/src/main/kotlin/info/papdt/express/helper/dao/PackageDatabase.kt
package info.papdt.express.helper.dao

import android.annotation.SuppressLint
import android.app.Application
import android.content.Context
import android.util.Log

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.annotations.Expose

import java.io.IOException
import java.util.ArrayList

import info.papdt.express.helper.api.PackageApi
import info.papdt.express.helper.api.PushApi
import info.papdt.express.helper.model.BaseMessage
import info.papdt.express.helper.model.Kuaidi100Package
import info.papdt.express.helper.support.FileUtils
import info.papdt.express.helper.support.SettingsInstance
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class PackageDatabase private constructor(private val mContext: Context) {

	@Expose @Volatile
	lateinit var data: ArrayList<Kuaidi100Package>
		private set
	@Expose
	var dataVersion: String? = "5.0.0"
		private set
	@Expose @Volatile
	var lastUpdateTime: Long = 0

	private var lastRemovedData: Kuaidi100Package? = null
	private var lastRemovedPosition = -1

	init {
		this.load()
	}

	fun load() {
		var json: String? = null
		try {
			json = FileUtils.readFile(mContext, FILE_NAME)
		} catch (e: IOException) {
			e.printStackTrace()
		}
        json = json ?: "{\"data\":[], \"lastUpdateTime\":0}"

		this.data = Gson().fromJson(json, PackageDatabase::class.java).data
	}

	fun restoreData(json: String) {
		this.data = Gson().fromJson(json, PackageDatabase::class.java).data
		CoroutineScope(Dispatchers.IO).launch { PushApi.sync(getPackageIdList()) }
	}

	val backupData: String
		get() {
			if (dataVersion == null) {
				dataVersion = "5.0.0"
			}
			return GsonBuilder().excludeFieldsWithoutExposeAnnotation().create().toJson(this)
		}

	fun save(): Boolean {
		return try {
			val gson = GsonBuilder().excludeFieldsWithoutExposeAnnotation().create()
			FileUtils.saveFile(mContext, FILE_NAME, gson.toJson(this))
			true
		} catch (e: IOException) {
			e.printStackTrace()
			false
		}
	}

    @Synchronized
	fun add(pack: Kuaidi100Package) {
		data.add(pack)
		CoroutineScope(Dispatchers.IO).launch { PushApi.add(pack.number!!, pack.companyType) }
	}

    @Synchronized
	fun add(index: Int, pack: Kuaidi100Package) {
		data.add(index, pack)
		CoroutineScope(Dispatchers.IO).launch { PushApi.add(pack.number!!, pack.companyType) }
	}

    @Synchronized
	operator fun set(index: Int, pack: Kuaidi100Package) {
		data[index] = pack
	}

    @Synchronized
	fun remove(index: Int) {
		val removedItem = data.removeAt(index)

		CoroutineScope(Dispatchers.IO).launch { PushApi.remove(removedItem.number!!) }

		lastRemovedData = removedItem
		lastRemovedPosition = index
	}

    @Synchronized
	fun remove(pack: Kuaidi100Package) {
		val nowPos = indexOf(pack)
		remove(nowPos)
	}

    @Synchronized
	fun undoLastRemoval(): Int {
		if (lastRemovedData != null) {
			val insertedPosition: Int = if (lastRemovedPosition >= 0 && lastRemovedPosition < data.size) {
				lastRemovedPosition
			} else {
				data.size
			}

			data.add(insertedPosition, lastRemovedData!!)

			lastRemovedData = null
			lastRemovedPosition = -1

			return insertedPosition
		} else {
			return -1
		}
	}

	fun indexOf(p: Kuaidi100Package): Int {
		return indexOf(p.number!!)
	}

	fun indexOf(number: String): Int {
		return (0 until size()).firstOrNull { get(it).number == number } ?: -1
	}

	fun clear() {
		data.clear()
	}

	fun size(): Int {
		return data.size
	}

	operator fun get(index: Int): Kuaidi100Package {
		return data[index]
	}

	fun getByNumber(number: String): Kuaidi100Package? {
		return data.find { it.number == number }
	}

	fun getPackageIdList(): List<String> = data.map { "${it.number}+${it.companyType}" }

	fun pullDataFromNetwork(shouldRefreshDelivered: Boolean) {
		if (SettingsInstance.enablePush) {
			CoroutineScope(Dispatchers.IO).launch { PushApi.sync(getPackageIdList()) }
		}
		for (i in 0 until size()) {
			val pack = this[i]
			if (!shouldRefreshDelivered && pack.getState() == Kuaidi100Package.STATUS_DELIVERED) {
				continue
			}
			val newPack = PackageApi.getPackage(pack.number!!, pack.companyType)
			if (newPack.code == BaseMessage.CODE_OKAY && newPack.data?.data != null) {
				pack.applyNewData(newPack.data)
				this[i] = pack
			} else {
				Log.e(TAG, "Kuaidi100Package " + pack.codeNumber + " couldn\'t get new info.")
			}
		}
	}

    fun readAll(): Int {
        var count = 0
		for (i in 0 until size()) {
			if (data[i].unreadNew) {
				count++
                data[i] = Kuaidi100Package(data[i])
				data[i].unreadNew = false
			}
		}
        return count
    }

    companion object {

		@SuppressLint("StaticFieldLeak")
        @Volatile
        private var sInstance: PackageDatabase? = null

		private const val FILE_NAME = "packages.json"

		private val TAG = PackageDatabase::class.java.simpleName

		fun getInstance(context: Context): PackageDatabase {
			if (sInstance == null) {
				synchronized(PackageDatabase::class.java) {
					if (sInstance == null) {
						sInstance = PackageDatabase(
                                if (context is Application) context else context.applicationContext
                        )
					}
				}
			}
			return sInstance!!
		}
	}

}
